python/perf-kernels/03-matrix-multiplication-stream-k.py

```diff
@@ -0,0 +1,407 @@
+# ---------------------------------------------------------------------------
+
+
+@triton.jit()
+def swizzle_tile(tile_id,
+                 M, N, K,
+                 BLOCK_M: tl.constexpr, BLOCK_N: tl.constexpr, BLOCK_K: tl.constexpr,
+                 GROUP_M: tl.constexpr
+                 ):
+    grid_m = tl.cdiv(M, BLOCK_M)
+    grid_n = tl.cdiv(N, BLOCK_N)
+    # re-order program ID for better L2 performance
+    width = GROUP_M * grid_n
+    group_id = tile_id // width
+    group_size = tl.minimum(grid_m - group_id * GROUP_M, GROUP_M)
+    pid_m = group_id * GROUP_M + (tile_id % group_size)
+    pid_n = (tile_id % width) // group_size
+    return pid_m, pid_n
+
+
+@triton.jit()
+def linear_tile(tile_id,
+                M, N, K,
+                BLOCK_M: tl.constexpr, BLOCK_N: tl.constexpr, BLOCK_K: tl.constexpr,
+                GROUP_M: tl.constexpr
+                ):
+    pid_m = tile_id // tl.cdiv(N, BLOCK_N)
+    pid_n = tile_id % tl.cdiv(N, BLOCK_N)
+    return pid_m, pid_n
+
+
+# iterate, multiply and accumulate over K axis
+@triton.jit()
+def mac_loop(A, B, C,
+             M, N, K,
+             locks,
+             stride_am, stride_ak, stride_bk, stride_bn, stride_cm, stride_cn,
+             iters_per_tile,
+             start_iter, end_iter,
+             BLOCK_M: tl.constexpr, BLOCK_N: tl.constexpr, BLOCK_K: tl.constexpr,
+             ACC_TYPE: tl.constexpr, GROUP_M: tl.constexpr):
+
+    # where are we in the grid
+    tile_id = start_iter // iters_per_tile
+    if GROUP_M  > 0:
+        pid_m, pid_n = swizzle_tile(tile_id, M, N, K, BLOCK_M, BLOCK_N, BLOCK_K, GROUP_M)
+    else:
+        pid_m, pid_n = linear_tile(tile_id, M, N, K, BLOCK_M, BLOCK_N, BLOCK_K, GROUP_M)
+
+    rm = pid_m * BLOCK_M + tl.arange(0, BLOCK_M)
+    rn = pid_n * BLOCK_N + tl.arange(0, BLOCK_N)
+    rk = tl.arange(0, BLOCK_K)
+    A = A + rm[:, None] * stride_am + rk[None, :] * stride_ak + BLOCK_K * stride_ak * (start_iter % iters_per_tile)
+    B = B + rk[:, None] * stride_bk + rn[None, :] * stride_bn + BLOCK_K * stride_bk * (start_iter % iters_per_tile)
+    acc = tl.zeros((BLOCK_M, BLOCK_N), dtype=ACC_TYPE)
+
+    for current_iter in range(start_iter, end_iter):
+        a = tl.load(A)
+        b = tl.load(B)
+        acc += tl.dot(a, b)
+        A += BLOCK_K * stride_ak
+        B += BLOCK_K * stride_bk
+
+    if end_iter % iters_per_tile == 0:  # last iteration of the tile always happens before its start on another SM
+        C_ = C + rm[:, None] * stride_cm + rn[None, :] * stride_cn  # compute inside the if/else to avoid spilling!
+        tl.store(C_, acc)
+        if start_iter % iters_per_tile != 0:  # only if tile has been partially processed
+            tl.atomic_xchg(locks + tile_id, 1)
+    else:
+        while tl.atomic_cas(locks + tile_id, 1, 1) != 1:
+            pass
+        C_ = C + rm[:, None] * stride_cm + rn[None, :] * stride_cn  # compute inside the if/else to avoid spilling!
+        tl.atomic_add(C_, acc)
```

Comment:
Any way we can avoid atomics in Triton? If we can allocate some additional memory for partials, this can all be done without the `atomic_add`, can be a lot more performant! Some notes from a colleague of mine that works on Tensile version are:

> For Tensile/rocBLAS we have implemented some kernels using atomics, but there are several factors that have made us limit their use or gradually remove kernels with atomics.
> 1. Atomics don't guarantee order of operations, and consistently reproducible results are essential for our HPC customers.
> 2. We sometimes use atomics as a stepping stone for either functional or high-performance kernels, but they gradually get phased out as we develop more optimizations so that non-atomic kernels outperform the atomic ones.
> 3. We typically start functional implementation of new hardware on emulators, and atomics are not always supported in emulation.
> 4. There are a number of hardware-specific issues to watch out for when using atomics:
>     - IIRC the register to control denorm settings for atomic ops doesn't have the same settings as the non-atomic variants
>     - For certain hardware/memory configurations (including MI200, gfx10/11), a number of atomic operations silently convert to a NOP instead
>     - For some add operations on MI100, the operation produces correct result, but does not match atomic API
>     - On MI300, some atomic operations had a scope downgrade - ie: device-scope rather than system-scope
>     - The work-around for these sorts of issues is to use a CAS loop, which may add a bit of overhead in some cases

Reply:
for reference, I have just comment out mac_loop, the timing is below:
(py_3.9) root@banff-cyxtera-s71-2:/home/work/triton/python/perf-kernels# python 03-matrix-multiplication-stream-kopt.py
total SMs: 304
M,N,K=8192,8192,8192 ; BLK_M,N,K=128,128,32
total_blocks_M=64 x total_blocks_N=64 = total_tiles=4096
total_tiles_streamk=448 + total_blocking_tiles=3648 = total_tiles=4096
total_programs_streamk=304
total_blocking_tiles=3648
iters_per_tile=256
total_iters_streamk=114688
0 registers used, 0 spills
0 registers used, 0 spills
PyTorch: 3.078 ms  357.269 tflops
hybrid stream-k (grid=304): 4.926 ms  223.208 tflops
hybrid stream-k (grid=608): 4.105 ms  267.861 tflops
tile matmul (grid=0): 5.582 ms  196.966 tflops

leave mac_loop, and comment out full_tile:
(py_3.9) root@banff-cyxtera-s71-2:/home/work/triton/python/perf-kernels# python 03-matrix-multiplication-stream-kopt.py
total SMs: 304
M,N,K=8192,8192,8192 ; BLK_M,N,K=128,128,32
total_blocks_M=64 x total_blocks_N=64 = total_tiles=4096
total_tiles_streamk=448 + total_blocking_tiles=3648 = total_tiles=4096
total_programs_streamk=304
total_blocking_tiles=3648
iters_per_tile=256
total_iters_streamk=114688
0 registers used, 0 spills
PyTorch: 3.064 ms  358.877 tflops
hybrid stream-k (grid=304): 0.744 ms  1477.277 tflops
hybrid stream-k (grid=608): 1.791 ms  613.757 tflops
tile matmul (grid=0): 0.042 ms  26012.623 tflops

so the major bottleneck is in full_tiles

Reply:
Which makes sense, I didn't imagine that the performance of atomics would dominate mac_loop, it was just meant to be an addition optimization to improve the overall throughput. Now, if the issue is within mac_loop, I can look at whats going on there, but do we have a comparison against data-parallel GEMM kernel (an existing kernel) to see what the impact of mac_loop in each is?